### DISCUSSION TITLE
Validation of configs using model information

### DISCUSSION THREAD

**Discussion by didiladi** [Dynatrace/dynatrace-configuration-as-code#209] [Ideas]
This is a summary of an internal discussion ticket, which we had some time ago. I'm adding it as a discussion ticket, feel free to continue the discussion where we last left off if you have some more input.

**Problem description**
We want to have more validation capabilities in `monaco` and to:
* Ensure that the config we are about to throw against the API is valid.
* We want to validatie the config *before* executing the web request.
* Have an option to validate against different API versions. E.g. you can use the same configuration in two environments with different versions. The validation should happen against the current API version of each individual enviroment. This helps us when there is a breaking API change.
* Give the users of `monaco` more feedback if the configuration is not valid. We want to be able to point to the non-valid json properties in the error message.

**Describe the solution you'd like**
During validation (`--dry-run`):
* Fetch API spec from the environment you want to apply the configs to
  * If the API doesn't exist (don't do anything for this configuration) => Needs dependency handling then afterwards
  * If the API exists, cross check the payload to the schema that we get from the spec and only deploy if they don't match up

@UnseenWizzard wrote:
> Swagger/OpenAPI follows json schema draft 4, but we need to check what version of swagger the cluster actually uses (http://json-schema.org/)
> There's a go validiator implementation of v4 (and 3 of the current v6 draft)
> [validate-json](https://github.com/cesanta/validate-json)  supports draft 4 (GPLv2)

**Comment by tatomir146** [comment]
@didiladi this way we would only check schema, but not if the upload actually works, which still is improvement but not a guarantee that upload is going to work. We could additionally use `validate` endpoint (if available) but this wouldn't work during `--dry-run` as if there are some dependencies defined to configuration not yet being uploaded.

**Comment by didiladi** [comment]
@tatomir146 yes, you are completely right. In the early days of the tooling I tried to use the `validate` endpoint, but I failed miserably with finding a generic solution. But I guess that would be worth another try since we learned quite a lot about our APIs in the meanwhile. On that note: Does settings 2.0 have the `validate` endpoint?
